homeassistant/components/esphomelib/config_flow.py

```diff
@@ -0,0 +1,129 @@
+    """Handle a esphomelib config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize flow."""
+        self._host = None  # type: Optional[str]
+        self._port = None  # type: Optional[int]
+        self._password = None  # type: Optional[str]
+        self._name = None  # type: Optional[str]
+
+    async def async_step_user(self, user_input: Optional[ConfigType] = None,
+                              error: Optional[str] = None):
+        """Handle a flow initialized by the user."""
+        if user_input is not None:
+            self._host = user_input['host']
+            self._port = user_input['port']
+            error, device_info = await fetch_device_info(self._host,
+                                                         self._port)
+            if error is not None:
+                return await self.async_step_user(error=error)
+            self._name = device_info.name
+
+            # Only show authentication step if device uses password
+            if device_info.uses_password:
+                return await self.async_step_authenticate()
+
+            return self._async_get_entry()
+
+        fields = OrderedDict()
+        fields[vol.Required('host', default=self._host or vol.UNDEFINED)] = str
+        fields[vol.Optional('port', default=self._port or 6053)] = int
+
+        errors = {}
+        if error is not None:
+            errors['base'] = error
+
+        return self.async_show_form(
+            step_id='user',
+            data_schema=vol.Schema(fields),
+            errors=errors
+        )
+
+    def _async_get_entry(self):
+        return self.async_create_entry(
+            title=self._name,
+            data={
+                'host': self._host,
+                'port': self._port,
+                # The API uses protobuf, so empty string denotes absence
+                'password': self._password or '',
+            }
+        )
+
+    async def async_step_authenticate(self, user_input=None, error=None):
+        """Handle getting password for authentication."""
+        if user_input is not None:
+            self._password = user_input['password']
+            error = await try_login(self._host, self._port, self._password)
+            if error:
+                return await self.async_step_authenticate(error=error)
+            return self._async_get_entry()
+
+        errors = {}
+        if error is not None:
+            errors['base'] = error
+
+        return self.async_show_form(
+            step_id='authenticate',
+            data_schema=vol.Schema({
+                vol.Required('password'): str
+            }),
+            errors=errors
+        )
+
+
+async def fetch_device_info(host, port):
+    """Fetch device info from API and return any errors."""
+    from aioesphomeapi.client import APIClient, APIConnectionError
+
+    cli = APIClient(asyncio.get_event_loop(), host, port, '')
```

Comment:
Shouldn't we use `hass.loop` explicitly here?

Reply:
Well a) this method will only ever be called from the HA event loop anyway. And b) if it would be called from another event loop, `asyncio.get_event_loop()` would be the correct behavior as otherwise this could result in a deadlock.

Reply:
We could pass in a loop parameter to the function, and the function still doesn't know what loop it is, but the caller knows what loop will be used.